src/core/components/files-regular/add-async-iterator.js

```diff
@@ -0,0 +1,200 @@
+'use strict'
+
+const importer = require('ipfs-unixfs-importer')
+const isSource = require('is-pull-stream').isSource
+const validateAddInput = require('ipfs-utils/src/files/add-input-validation')
+const { parseChunkerString } = require('./utils')
+const pipe = require('it-pipe')
+const { supportsFileReader } = require('ipfs-utils/src/supports')
+const isStream = require('is-stream')
+const toAsyncIterator = require('pull-stream-to-async-iterator')
+const log = require('debug')('ipfs:add')
+log.error = require('debug')('ipfs:add:error')
+
+function noop () {}
+
+module.exports = function (self) {
+  // Internal add func that gets used by all add funcs
+  return async function * addAsyncIterator (source, options) {
+    options = options || {}
+
+    let chunkerOptions = parseChunkerString(options.chunker)
+
+    const opts = Object.assign({}, {
+      shardSplitThreshold: self._options.EXPERIMENTAL.sharding
+        ? 1000
+        : Infinity
+    }, options, {
+      chunker: chunkerOptions.chunker,
+      chunkerOptions: chunkerOptions.chunkerOptions
+    })
+
+    // CID v0 is for multihashes encoded with sha2-256
+    if (opts.hashAlg && opts.cidVersion !== 1) {
+      opts.cidVersion = 1
+    }
+
+    let total = 0
+
+    const prog = opts.progress || noop
+    const progress = (bytes) => {
+      total += bytes
+      prog(total)
+    }
+
+    opts.progress = progress
+
+    if (Buffer.isBuffer(source) || typeof source === 'string') {
+      source = [
+        source
+      ]
+    }
+
+    const iterator = pipe(
+      source,
+      validateInput(),
+      normalizeInput(opts),
```

Comment:
I made a comprehensive normalise in https://github.com/ipfs/js-ipfs-http-client/pull/1063/files#diff-7cec0363d7e0ecfb1f2d5e04b55e2cac and https://github.com/ipfs/js-ipfs-http-client/pull/1063/files#diff-8913549ebf1a95e3151ad16513fa9a6f - it would be good to check this is supporting the same inputs (or come to an agreement on what inputs we support).

Reply:
Nice, maybe we could move that into [`ipfs-utils`](https://github.com/ipfs/js-ipfs-utils)?

Reply:
Yes, will do 👍

Reply:
I'm going to have a go at moving those utility functions over to `ipfs-utils` so they can be shared between the http client & core as it's going to simplify this a little.

Reply:
I'm going to refactor some of it though, because it's full of `Buffer.from` which involves a buffer copy which is slow.